csharp/src/Ice/CurrentExtensions.cs

```diff
@@ -113,25 +113,29 @@ public static OutgoingResponse createOutgoingResponse(this Current current, bool
     /// <param name="current">The current object of the corresponding incoming request.</param>
     /// <param name="exception">The exception to marshal into the response payload.</param>
     /// <returns>A new outgoing response.</returns>
-    public static OutgoingResponse createOutgoingResponse(this Current current, System.Exception exception)
+    public static OutgoingResponse createOutgoingResponse(
+        this Current current,
+        System.Exception exception,
+        Communicator? communicator = null)
     {
+        communicator ??= current.adapter.getCommunicator() ?? throw new ArgumentNullException(nameof(communicator));
```

Comment:
My understanding is the bug you want to fix is current.adapter is occasionally null.

I'd rather not leak this special case outside ConnectionI; in particular, I want to keep current.adapter as non-nullable.
And if you wanted to do that, the correct syntax is actually `current.adapter?.getCommunicator()`.

Reply:
Yes that is the bug. 

I think it is bad to have the IncomingRequest creating a current object with a null adapter.

Either we can make IncomingRequest.current nullable, or move the reading of the request header to a helper method, and don't create an IncomingRequest object when there is no adapter.

Reply:
> I think it is bad to have the IncomingRequest creating a current object with a null adapter.

I don't see a good alternative.

If we create Current with a Current constructor that takes an InputStream and an ObjectAdapter (and more), we'll hit the same issue: the ObjectAdapter is null in the rare situation where you receive a request but there is no OA, but we don't want Current.adapter to be nullable.